client/src/utils/ajax.ts

```diff
@@ -0,0 +1,164 @@
+import envData from '../../../config/env.json';
+import Tokens from 'csrf';
+import cookies from 'browser-cookies';
+
+import type { UserType } from '../redux/prop-types';
+
+const { apiLocation } = envData as { apiLocation: string };
+
+const base = apiLocation;
+const tokens = new Tokens();
+
+const defaultOptions: RequestInit = {
+  credentials: 'include'
+};
+
+// _csrf is passed to the client as a cookie. Tokens are sent back to the server
+// via headers:
+function getCSRFToken() {
+  const _csrf = typeof window !== 'undefined' && cookies.get('_csrf');
+  if (!_csrf) {
+    return '';
+  } else {
+    return tokens.create(_csrf);
+  }
+}
+
+async function get<T>(path: string): Promise<T> {
+  return fetch(`${base}${path}`, defaultOptions).then(res =>
+    res.json()
+  ) as Promise<T>;
+}
+
+export function post<T = void>(path: string, body: unknown): Promise<T> {
+  return request('POST', path, body);
+}
+
+function put<T = void>(path: string, body: unknown): Promise<T> {
+  return request('PUT', path, body);
+}
+
+async function request<T>(
```

Comment:
I do not see this calltime ( _not correct terminology_ ) type `T` being used anywhere.

Reply:
Not explicitly, but TS correctly deduces its implicit use in `post()` and `put()`. It can be explicitly used if that's preferred, should I do that?

Before:
```js
export function post<T = void>(path: string, body: unknown): Promise<T> {
  return request('POST', path, body);
}

function put<T = void>(path: string, body: unknown): Promise<T> {
  return request('PUT', path, body);
}
```

After:
```js
export function post<T = void>(path: string, body: unknown): Promise<T> {
  return request<T>('POST', path, body);
}

function put<T = void>(path: string, body: unknown): Promise<T> {
  return request<T>('PUT', path, body);
}
```

Reply:
On the subject of generics, I've just discovered that `.then()` accepts a generic type, which removes the need to assert `as Promise<T>`.

Reply:
> It can be explicitly used if that's preferred, should I do that?

No, I was just unaware of TS being able to infer such.